internal/moduletest/graph/node_provider.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package graph
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hcldec"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/hashicorp/terraform/internal/moduletest"
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+	"github.com/hashicorp/terraform/version"
+)
+
+var (
+	_ GraphNodeExecutable = (*NodeProviderConfigure)(nil)
+	_ GraphNodeReferencer = (*NodeProviderConfigure)(nil)
+)
+
+type NodeProviderConfigure struct {
+	name, alias string
+
+	Addr     addrs.RootProviderConfig
+	File     *moduletest.File
+	Config   *configs.Provider
+	Provider providers.Interface
+}
+
+func (n *NodeProviderConfigure) Name() string {
+	if len(n.alias) > 0 {
+		return fmt.Sprintf("provider.%s.%s", n.name, n.alias)
+	}
+	return fmt.Sprintf("provider.%s", n.name)
+}
+
+func (n *NodeProviderConfigure) Execute(ctx *EvalContext) {
+	log.Printf("[TRACE]: NodeProviderConfigure: configuring provider %s for tests", n.Addr)
+	if ctx.Cancelled() || ctx.Stopped() {
+		return
+	}
+
+	// first, set the provider so everything else can use it
+	ctx.SetProvider(n.Addr, n.Provider)
+
+	spec := n.Provider.GetProviderSchema().Provider.Body.DecoderSpec()
```

Comment:
nit: we could get and store the provider schema on the struct during transformation, and not have to call it multiple times. The main terraform graph also tries to do something similar by caching, but we dont have access to that cache here.